src/rendering/MaterialParser/MaterialParser.cc

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (C) 2023 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <cstddef>
+#include <iostream>
+#include <map>
+#include <string>
+#include <vector>
+
+#include "MaterialParser.hh"
+
+using namespace gz;
+using namespace sim;
+
+/////////////////////////////////////////////////
+MaterialParser::MaterialParser()
+  : configLoader()
+{
+}
+
+/////////////////////////////////////////////////
+void MaterialParser::Load()
+{
+  ConfigLoader::loadMaterialFiles(&this->configLoader);
+}
+
+/////////////////////////////////////////////////
+std::optional<MaterialParser::MaterialValues> MaterialParser::GetMaterialValues(
+  const std::string& material)
+{
+  std::optional<MaterialValues> values = std::nullopt;
+  std::map<std::string, ConfigNode *> scripts =
+    this->configLoader.getAllConfigScripts();
+
+  std::map<std::string, ConfigNode *> ::iterator it;
+
+  for (it = scripts.begin(); it != scripts.end(); ++it) {
+    std::string name = it->first;
+    if (name.find(material) != std::string::npos) {
+      if(!values) {
+        values = MaterialValues();
+      }
+      ConfigNode * node = it->second;
+
+      ConfigNode * techniqueNode = node->findChild("technique");
+      if (techniqueNode) {
+        ConfigNode * passNode = techniqueNode->findChild("pass");
+        if (passNode) {
+          ConfigNode * ambientNode = passNode->findChild("ambient");
+          if (ambientNode) {
+            std::vector<float> ambientValues;
+            ambientNode->getValuesInFloat(ambientValues);
```

Comment:
We can't guarantee that `getValuesInFloat` will return a vector of 3 (at least). For example, if the user modifies the `gazebo.material` file and forgets to put the third value for a color, we'd end up with a vector of size 2, but then we'd try to access `ambientValues[2]` which would be undefined behavior. So I think we should add a check here and other places `getValuesInFloat` is used in this file. Alternatively, we can have a fuction in `ConfigNode` that returns a color instead of a vector.

Reply:
Instead of `gz/math` dependency in `ConfigLoader` and placing extra checks in `MaterialParser`, I'm thinking of updating `getValuesInFloat` to check if the vector has 4 or more values, which is common for all colors in the material file. If not then throwing an error saying "Bad material file", what do you think about that?

Reply:
If we do that, my preference would be to create another `getColorValues` function or something similar so `getValuesInFloat` can remain a generic function that gets a list of floats  for  non-colors. Either way, by "throwing an error",  you mean printing an error message, I definitely agree. But if you mean throw an exception, I would say we should avoid throwing exceptions since that's our general practice in the Gazebo codebase.

Reply:
Added another function with the size check. Yeah that's right just printing an error message, not throwing an exception.

Reply:
Okay, I see the size check has been added, but it would still allow out of bounds access on the vector since the code continues to assign the colors after `getColorValues` whether that fails or not. I would suggest something like the following:
```c++
            std::vector<float> ambientValues;
            ambientNode->getColorValues(ambientValues);
            values->ambient.emplace();
            for (std::size_t i=0; i < ambientValues.size(); ++i)
            {
                   values->ambient[i] = ambientValues[i];
            }
```

However, it looks like there's a bug in `Color::operator[]` since it doesn't return a reference. So another way would be to modify the `getColorValues` function to resize the vector.
```c++
  inline void getColorValues(std::vector<float> & colorValues)
  {
    getValuesInFloat(colorValues);
    if (colorValues.size() < 3) 
   {
      gzerr << "Bad material file." << std::endl;
      colorValues.resize(3);
    }
  }
```